src/Microsoft.AspNet.OData.Shared/Query/Expressions/FilterBinder.cs

```diff
@@ -329,16 +329,16 @@ private Expression BindSingleResourceCastFunctionCall(SingleResourceFunctionCall
                 targetClrType = EdmLibHelpers.GetClrType(targetEdmType.ToEdmTypeReference(false), Model);
             }
 
-            if (arguments[0].Type == targetClrType)
+            Expression source = arguments[0];
+
+            if (source.Type == targetClrType)
             {
                 // We only support to cast Entity type to the same type now.
-                return arguments[0];
+                return source;
             }
-            else if (arguments[0].Type.IsAssignableFrom(targetClrType))
+            else if (source.Type.IsAssignableFrom(targetClrType))
             {
                 // To support to cast Entity/Complex type to the sub type now.
-                Expression source = BindCastSourceNode(node.Source);
```

Comment:
>Expression source = BindCastSourceNode(node.Source); [](start = 16, length = 52)

why remove this line?

Reply:
@xuzhg 
`BindCastSourceNode` binds the `Source` of either the `SingleResourceCastNode` or the `CollectionResourceCastNode`.

If the cast is on the root e.g 
`/Products?$filter=cast('Microsoft.AspNet.OData.Test.Query.Expressions.DerivedProduct')/DerivedCategoryName eq null` , the `Source` will always be null hence `BindCastSourceNode` will always return a parameterExpression with the expression `$it`.
In a scenario where we are casting a navigation property to return an expression that queries against the parent property,
we need to have a `memberAccess` expression. e.g `$it.Category` in the `lambdaExpression`

In the example below:
`/Products?$filter=cast(Category,'Microsoft.AspNet.OData.Test.Query.Expressions.DerivedCategory')/DerivedCategoryName eq null`
should return the lambda expression `$it => (($it.Category As DerivedCategory).DerivedCategoryName == null)`.
However if we call `BindCastSourceNode`, we will get
`$it => (($it As DerivedCategory).DerivedCategoryName == null)`

I have modified the code to only Call `BindCastSourceNode` when not casting on the root.

cc @mikepizzo @gathogojr @Sreejithpin